othello/apps/games/consumers.py

```diff
@@ -0,0 +1,106 @@
+
+
+class GameConsumer(JsonWebsocketConsumer):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.game = None
+        self.connected = False
+
+    def connect(self):
+        try:
+            self.game = Game.objects.get(id=self.scope["url_route"]["kwargs"]["game_id"])
+        except Game.DoesNotExist:
+            self.close()
+            return
+
+        yourself = Submission.objects.get(user__username="Yourself")
+
+        if not self.game.playing:
+            self.close()
+            return
+
+        self.is_black_yourself = self.game.black == yourself
+        self.is_white_yourself = self.game.white == yourself
+
+        self.connected = True
+        self.accept()
+
+        async_to_sync(self.channel_layer.group_add)(
+            self.game.channels_group_name, self.channel_name
+        )
+
+    def disconnect(self, code):
+        self.connected = False
+        super().disconnect(code=code)
+        self.close()
+
+    def game_update(self, event):
+        self.update_game()
+
+    def game_log(self, event):
+        self.send_log(event["object_id"])
+
+    def game_error(self, event):
+        self.send_error(event["object_id"])
+
+    def game_ping(self, event):
+        self.send_json({"type": "game.ping"})
+
+    def update_game(self):
+        if self.connected:
+            self.game.refresh_from_db()
+            game = serialize_game_info(self.game)
+            self.send_json(game)
+            if game["game_over"]:
+                self.disconnect(code=0)
+
+    def send_log(self, object_id):
+        if self.connected:
+            self.game.refresh_from_db()
+            log = self.game.logs.get(id=object_id)
+            has_access = (
+                log.game.black.user == self.scope["user"]
+                if log.player == Player.BLACK.value
+                else log.game.white.user == self.scope["user"]
+            )
+            if (
+                has_access
+                or (log.player == Player.BLACK.value and self.is_black_yourself)
+                or (log.player == Player.WHITE.value and self.is_white_yourself)
+            ):
+                self.send_json(serialize_game_log(log))
+
+    def send_error(self, object_id):
+        if self.connected:
+            self.send_json(serialize_game_error(self.game.errors.get(id=object_id)))
+
+
+class GamePlayingConsumer(GameConsumer):
+    def connect(self):
+        super(GamePlayingConsumer, self).connect()
+        run_game.delay(self.game.id)
+
+    def disconnect(self, code):
+        self.game.playing = False
+        self.game.save(update_fields=["playing"])
+        super(GamePlayingConsumer, self).disconnect(code)
+
+    def receive_json(self, content, **kwargs):
+        self.game.last_heartbeat = timezone.now()
+        self.game.save()
+
+        player = content.get("player", False)
+        if move := int(content.get("move", False)):
```

Comment:
Why aren't you using `None`s here? WiIl `move` ever be 0? If so, you're in trouble.

This needs to be refactored.